### PR TITLE
`Development`: Move logic for saving assessments into a service

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
While working on #7552, I noticed that the `AssessmentResource` and `ProgrammingAssessmentResource` contained logic for saving and submitting assessments, but this logic should be part of the corresponding assessment services.

### Description
Moved the code from the resource to the service.

This PR also fixes a small bug when publishing assessments via LTI. We now only publish the result when the assessment gets submitted, not when it's just saved. Similarly, manual feedback requests are only resolved once the assessment got submitted and not just saved.

### Steps for Testing
code review

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2

### Test Coverage
unchanged

## Summary by CodeRabbit

- **New Features**
  - Streamlined submission process for non-programming exercises with simplified due date handling.
  - Enhanced assessment services with real-time result updates and user notifications.

- **Improvements**
  - Constructor modifications to include new services for user notifications and result updates.
  - Refactored assessment methods for manual submissions to improve efficiency.

- **Bug Fixes**
  - Removed unnecessary due date parameter in result submissions for cleaner and more consistent functionality.

- **Refactor**
  - Eliminated unused imports and dependencies across various service and resource files to clean up the codebase.
  - Adjusted constructors and methods for better alignment with new assessment processes.

- **Tests**
  - Updated test cases to reflect changes in result submission and assessment service logic.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 1]
```diff
@@ -268,6 +275,33 @@ public Result saveManualAssessment(final Submission submission, final List<Feedb
...
+     * @param exercise     the exercise this assessment belongs to
+     * @param submission   the assessed submission
+     * @param feedbackList the assessment as a feedback list that should be added to the result of the corresponding submission
+     * @param resultId     if of the result we what to save the feedbackList to, null if no result exists
+     * @param submit       true if the result should also be submitted
+     * @return the saved result
+     */
+    public Result saveAndSubmitManualAssessment(final Exercise exercise, final Submission submission, final List<Feedback> feedbackList, Long resultId, boolean submit) {
+        Result result = saveManualAssessment(submission, feedbackList, resultId);
+        if (submit) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:291]: you could return early here.

**Review by MaximilianAnzinger — DISMISSED** [review]
Code looks good 👍 I left one optional suggestion

**Review by basak-akan — APPROVED** [review]
Code looks good. I also tested lti Moodle grade updating scenarios. Both manual and automatic grading results are correctly updated in Moodle.

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 2]
```diff
@@ -229,12 +236,12 @@ public Result submitManualAssessment(long resultId, Exercise exercise) {
...
      * <p>
-     * For programming exercises we use a different approach see {@link ProgrammingAssessmentService#saveManualAssessment(Result)}
+     * For programming exercises we use a different approach see {@link ProgrammingAssessmentService#saveManualAssessment(Result, User)}.
      *
-     * @param submission   the file upload submission to which the feedback belongs to
+     * @param submission   the submission to which the feedback belongs to
      * @param feedbackList the assessment as a feedback list that should be added to the result of the corresponding submission
-     * @param resultId     resultId of the submission we what to save the @feedbackList to, null if no result exists
-     * @return result that was saved in the database
+     * @param resultId     if of the result we what to save the feedbackList to, null if no result exists
```
Comment (chrisknedl) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:243]:
```suggestion
     * @param resultId     if the result we want to save the feedbackList to exists, null if not
```

**Review by chrisknedl — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 3]
```diff
@@ -229,12 +236,12 @@ public Result submitManualAssessment(long resultId, Exercise exercise) {
...
      * <p>
-     * For programming exercises we use a different approach see {@link ProgrammingAssessmentService#saveManualAssessment(Result)}
+     * For programming exercises we use a different approach see {@link ProgrammingAssessmentService#saveManualAssessment(Result, User)}.
      *
-     * @param submission   the file upload submission to which the feedback belongs to
+     * @param submission   the submission to which the feedback belongs to
      * @param feedbackList the assessment as a feedback list that should be added to the result of the corresponding submission
-     * @param resultId     resultId of the submission we what to save the @feedbackList to, null if no result exists
-     * @return result that was saved in the database
+     * @param resultId     id of the result we what to save the feedbackList to, null if no result exists
```
Comment (chrisknedl) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:243]: Still needs to be "want", right?
```suggestion
     * @param resultId     id of the result we want to save the feedbackList to, null if no result exists
```

**Review by chrisknedl — DISMISSED** [review]
Code lgtm

**Review by tobias-lippert — APPROVED** [review]
Code

**Review by MaximilianAnzinger — DISMISSED** [review]

**Comment by Strohgelaender** [comment]
@coderabbitai review

File: src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingAssessmentIntegrationTest.java [hunk 4]
```diff
@@ -899,7 +899,9 @@ void unlockFeedbackRequestAfterAssessment() throws Exception {
 
         doNothing().when(programmingExerciseParticipationService).unlockStudentRepositoryAndParticipation(participation);
 
-        var response = request.putWithResponseBody("/api/participations/" + participation.getId() + "/manual-results", result, Result.class, HttpStatus.OK);
+        var params = new LinkedMultiValueMap<String, String>();
+        params.add("submit", "true");
+        var response = request.putWithResponseBodyAndParams("/api/participations/" + participation.getId() + "/manual-results", result, Result.class, HttpStatus.OK, params);
 
         var responseParticipation = response.getParticipation();
         assertThat(responseParticipation.getIndividualDueDate()).isNull();
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingAssessmentIntegrationTest.java]: I fixed that in eb87856f69d6f0c6da924d9a95c895540ec6f010.

**Review by basak-akan — APPROVED** [review]
Re-tested for LTI scenarios still works as expected. Recent changes lgtm.

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 5]
```diff
@@ -268,6 +275,35 @@ public Result saveManualAssessment(final Submission submission, final List<Feedb
...
     }
 
+    /**
+     * Saves a new manual assessment. Submits the result if the submit-parameter is set to true.
+     * Also notifies the student about the assessment if it is visible (after the assessment due date).
+     *
+     * @param exercise     the exercise this assessment belongs to
+     * @param submission   the assessed submission
+     * @param feedbackList the assessment as a feedback list that should be added to the result of the corresponding submission
+     * @param resultId     if of the result we what to save the feedbackList to, null if no result exists
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:285]:
```suggestion
     * @param resultId     if of the result we want to save the feedbackList to, null if no result exists
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingAssessmentService.java [hunk 6]
```diff
@@ -43,4 +57,85 @@ public Result saveManualAssessment(Result result, User assessor) {
...
+        if (athenaFeedbackSendingService.isPresent() && exercise.getFeedbackSuggestionsEnabled()) {
+            athenaFeedbackSendingService.get().sendFeedback(exercise, programmingSubmission, feedbacks);
+        }
+    }
+
+    private void handleResolvedFeedbackRequest(StudentParticipation participation) {
+        var exercise = participation.getExercise();
+        var isManualFeedbackRequest = exercise.getAllowManualFeedbackRequests() && participation.getIndividualDueDate() != null
+                && participation.getIndividualDueDate().isBefore(ZonedDateTime.now());
+        var isBeforeDueDate = exercise.getDueDate() != null && exercise.getDueDate().isAfter(ZonedDateTime.now());
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingAssessmentService.java:131]: is it possible to use a helper function from `ExerciseDateService`? maybe `isBeforeEarliestDueDate`?

**Review by reschandreas — DISMISSED** [review]
overall code lgtm

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingAssessmentService.java (hunk 6 above)
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingAssessmentService.java:135]: No, we need to check the exercise due date here and explicitly want to ignore the individual due date of the participation. I  added a comment to explain that.

**Review by reschandreas — APPROVED** [review]
code lgtm

**Review by aplr — APPROVED** [review]
Code LGTM

**Review by coolchock — APPROVED** [review]
code looks good to me

**Review by dearjasmina — APPROVED** [review]
code lgtm

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**